#1 Pull Request

EwertonRosendo opened the pull request.

EwertonRosendo committed:
added controller for books

EwertonRosendo committed:
added book_params, set_book and before_action

EwertonRosendo committed:
added edit and destroy actions

EwertonRosendo committed:
using model Author and Book

EwertonRosendo committed:
keeping the house clean, or not

EwertonRosendo committed:
added a new view: book by id

EwertonRosendo committed:
added update function

EwertonRosendo committed:
added delete function

EwertonRosendo commented:
![Screenshot from 2024-06-16 17-42-02](https://github.com/EwertonRosendo/books_api/assets/93278917/5aed8165-9476-4041-89e2-c44eda154ce8)

-> app/Books/id

It's the page to change or delete a book

EwertonRosendo committed:
small changes

EwertonRosendo committed:
Added field url_image for Book

EwertonRosendo committed:
CreateBook.js created to send new data

EwertonRosendo committed:
created handleAddGoogleBook

EwertonRosendo committed:
added  seach by title

EwertonRosendo committed:
added .order(published_at desc)

EwertonRosendo committed:
if the google api does not return any books, a different material will be shown

EwertonRosendo committed:
new we can change the author

EwertonRosendo committed:
css changes

EwertonRosendo committed:
added csrf tokens at requests

EwertonRosendo committed:
some css

EwertonRosendo commented:
/app -> books from google api
![image](https://github.com/EwertonRosendo/books_api/assets/93278917/b3b41936-e27b-4386-a9f2-48361fcd6888)
/app/Books -> books from Book Model
![image](https://github.com/EwertonRosendo/books_api/assets/93278917/ca33d6ce-e867-49cc-9b84-a1f330fff9f9)
/app/Books/id 
![image](https://github.com/EwertonRosendo/books_api/assets/93278917/cc5b4ec5-2e57-4ea5-9da2-0baa20fa3826)

EwertonRosendo committed:
pagination added at app/Books

EwertonRosendo committed:
RuboCop added

EwertonRosendo committed:
added graphql

EwertonRosendo committed:
added create author using graphql

EwertonRosendo committed:
now we can create books and authors using graphql

EwertonRosendo committed:
added user

EwertonRosendo committed:
back on rails

matcheus commented on app/controllers/app_controller.rb:
```diff
@@ -0,0 +1,64 @@
+require 'json'
+
+class AppController < ApplicationController
+  def main; end
```
O que é essa método? Essa rota é usada para o que? O nome dos métodos devem ter significado.

matcheus commented on app/controllers/app_controller.rb:
```diff
@@ -0,0 +1,64 @@
+
+  def books; end
+
+  def books_by_id; end
```
Esses métodos parecem deslocados, books dentro do app_controller?

matcheus commented on app/controllers/app_controller.rb:
```diff
@@ -0,0 +1,64 @@
+  def books_by_id; end
+
+  def index_json
+    title = params[:title] || 'Harry%20Potter'
```
Sua aplicação tem que utilizar uma forma default mais apropriada para quando nenhum titulo é passado, porque esse valor é por padrão em caso de `nil` setado pra `'Harry%20Potter'`?

matcheus commented on app/controllers/app_controller.rb:
```diff
@@ -0,0 +1,64 @@
+  def index_json
+    title = params[:title] || 'Harry%20Potter'
+
+    puts title
```
O que é esse `puts` no meio do código? É para debug? 
Nenhum `puts` ou qualquer outra ferramenta de debug podem entrar no código final, use só no ambiente local.

matcheus commented on app/controllers/app_controller.rb:
```diff
@@ -0,0 +1,64 @@
+      return render json: list_books
+    end
+
+    puts('no BOOKS')
```
Ditto sobre `puts`

matcheus commented on app/controllers/app_controller.rb:
```diff
@@ -0,0 +1,64 @@
+        list_books.push(book)
+      end
+      return render json: list_books
+    end
```
Toda lógica relacionada a requisições externas devem ser mantidas em um serviço fora do controller. A responsabilidade do controller é receber a requisição e retornar a resposta processada. Ele pode no máximo processar a entrada assim como formatar a response caso não seja uma lógica complexa.

matcheus commented on app/controllers/author_controller.rb:
```diff
@@ -0,0 +1,15 @@
+
+require 'json'
+class AuthorController < ApplicationController
+  protect_from_forgery with: :null_session
```
Não há necessidade de adicionar essa linha novamente já que você ja colocou ela no seu `ApplicationController` e essa classe herda do mesmo.

matcheus commented on app/controllers/books_controller.rb:
```diff
@@ -0,0 +1,66 @@
+require 'net/http'
+
+class BooksController < ApplicationController
+  # protect_from_forgery with: :null_session
```
Ditto

matcheus commented on app/controllers/author_controller.rb:
```diff
@@ -0,0 +1,15 @@
+  end
+
+  def show
+    author = Author.find_by(id: params[:id])
```
`find_by` é mais apropriado quando buscando por atributos alem do `id`, use apenas `find` se a busca é só pelo `id`.

matcheus commented on app/controllers/books_controller.rb:
```diff
@@ -0,0 +1,66 @@
+  def create
+    author = Author.find_or_create_author(params[:author])
+
+    book = {
+      title: params[:title],
+      description: params[:description],
+      published_at: params[:published_at],
+      publisher: params[:publisher],
+      url_image: params[:url_image],
+      author:
+    }
+    
+    render json: Book.create(book)
```
```
@book = Book.create(params)
```

Reforço a necessidade da leitura da [documentação oficial do active record](https://guides.rubyonrails.org/active_record_basics.html) para entender o funcionamento desses métodos e adiciono também a leitura da [documentação sobre os controllers](https://guides.rubyonrails.org/action_controller_overview.html), alem disso deve-se fazer a leitura tanto do [Ruby Style Guide](https://github.com/rubocop/ruby-style-guide) quanto do [Rails Style Guide](https://github.com/rubocop/rails-style-guide) em caráter EMERGENCIAL 🚨 .

matcheus reviewed (changes requested):
Revisar todo o código seguindo os seguintes documentos:
[Ruby Style Guide](https://github.com/rubocop/ruby-style-guide)
[Rails Style Guide](https://github.com/rubocop/rails-style-guide)

matcheus commented on app/controllers/books_controller.rb:
```diff
@@ -0,0 +1,66 @@
+    @book[:publisher] = params[:publisher]
+    @book[:author_id] = author.id
+    @book[:updated_at] = Time.current
+    @book[:url_image] = params[:url_image]
```
Essa é a sintaxe utilizada para Hash, outros objetos devem usar a seguinte sintaxe:

```
@book.title = 
@book.description = 
...
```

EwertonRosendo committed:
using respond_to with formats

EwertonRosendo committed:
removed files

EwertonRosendo committed:
created folder app/services with search_google_api and changed the controllers

EwertonRosendo committed:
create Book now is working better

EwertonRosendo committed:
removed find_or_create from Author's Model

EwertonRosendo committed:
changing some files

abxms commented:
Acho legal manter um padrão com o comentário dos commits. Geralmente, sempre usamos o imperativo, então não colocar no passado ou no gerúndio, mais pra manter um padrão mesmo. 

Exemplo: 
[using respond_to with formats](https://github.com/EwertonRosendo/books_api/pull/1/commits/be903a38a36e737adbc8e4e893a3214c6f499ffc) -> `use respond_to with formats`

Além de sempre atentar pra descrever bem quais as principais mudanças que tão sendo feitas em cada commit (assim fica mais fácil de encontrar uma versão, caso a gente queira dar revert em algo, ou encontrar quando mudamos alguma coisa específica)

Exemplo:
[removed files](https://github.com/EwertonRosendo/books_api/pull/1/commits/c36bb74bbbc5ed1702844ad6d61b3d9269c482c8) -> `remove views that were not being used`

(Considera isso quando for escrever os próximos, mas não precisa editar os que já existem. Também acho que não tem necessidade de usar _conventional commits_, só padronizar e descrever eles melhores mesmo)

[How to Write Better Git Commit Messages – A Step-By-Step Guide](https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/)

EwertonRosendo committed:
changes on search service and controllers

EwertonRosendo committed:
Update code syntax following rails-style-guide

Refactor code syntax to adhere to Rails Style Guide recommendations.

EwertonRosendo committed:
change controller and view app to GoogleBooks controller and view

EwertonRosendo committed:
create component ListBook and SearchBook

Create ListBook Component to show a list of books and SearchBook to send requests

EwertonRosendo committed:
remove blank lines from css files

gprado-justworks commented on app/javascript/components/MyBooks/CreateBook.js:
```diff
@@ -0,0 +1,75 @@
+
+const CreateBook = (props) => {
+
+  const [title, setTitle] = useState("")
```
tem varios states e todos eles sao parte do form, ficaria melhor deixar apenas um objeto que armazene tudo

Ex:
```js
const [formData, setFormData] = useSate();
```

Ai pra setar os valores:
```js
function handleInputChange(event) {
  const { id, value } = event.target;
 
  setFormData({
    ...formData,
    [id]: value,
  })
}
```

E trocar o elemento para:
```html
<input id="title" type="text" onChange={handleInputChange} placeholder={"Book title.."} />
```

msantos-justworks commented on app/helpers/app_helper.rb:
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+module AppHelper
```
Se não esta sendo utilizado, pode ser removido

msantos-justworks commented on app/helpers/author_helper.rb:
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+module AuthorHelper
```
Ditto

msantos-justworks commented on app/helpers/books_helper.rb:
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+module BooksHelper
```
Ditto

msantos-justworks commented on app/helpers/users_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
```
Ditto

EwertonRosendo committed:
remove files that are not being used and create a form in the CreateBooks.js file

gprado-justworks commented on app/javascript/components/MyBooks/CreateBook.js:
```diff
@@ -45,31 +42,31 @@ const CreateBook = (props) => {
         <div className="descrip">
             <label>Description:</label>
-            <textarea className="description" type="text" onChange={(e) => {setDescription(e.target.value)}} placeholder={"Book description.."} name="" id=""></textarea>
+            <textarea id="url_image" className="description" type="text" onChange={handleInputChange} placeholder={"Book description.."} name="" ></textarea>
```
Aqui ficou com o mesmo ID do input de cima, `url_image`

gprado-justworks commented on app/javascript/components/MyBooks/CreateBook.js:
```diff
@@ -3,13 +3,15 @@ import axios from "axios";
+    const {id, value} = event.target;
+    setFormData({
+      ...formData,
+      [id]:value
```
non-blocking: coloca um espacinho ` ` depois dos `:`, `[id]: value`

EwertonRosendo committed:
change input url_image to description

change input id to description and add a blank space on formData
